kms-message/src/kms_request.c

```diff
@@ -891,14 +898,19 @@ kms_request_free_string (char *ptr)
    free (ptr);
 }
 
-uint8_t *
+const uint8_t *
 kms_request_to_bytes (kms_request_t *request, uint32_t *len)
 {
-   if (request->provider != KMS_REQUEST_PROVIDER_KMIP) {
-      KMS_ERROR (request, "kms_request_to_bytes only applies to KMIP requests");
+   if (request->provider == KMS_REQUEST_PROVIDER_KMIP) {
+      *len = request->kmip.len;
+      return request->kmip.data;
+   }
+
+   if (!request->to_string && !kms_request_to_string (request)) {
       return NULL;
    }
 
-   *len = request->kmip.len;
-   return request->kmip.data;
+   KMS_ASSERT (request->to_string);
+   *len = (uint32_t) request->to_string->len;
+   return (const uint8_t*) request->to_string->str;
```

Comment:
> 1. Should the return type be `const char*`?

I think not. I think `uint8_t` conveys the intent of returning a byte rather than a value that should may be interpreted as a character. 

I was treating `uint8_t*` and `char*` as if they are interchangeable. `uint8_t` is guaranteed to be 8 bits. `char` is guaranteed to be one byte. After reading https://stackoverflow.com/a/1725901/774658 I see the size of `char` may exceed 8 bits on platforms where `CHAR_BITS` > 8. I do not think that applies to any platforms we are testing. So I think it is OK to make the assumption that `char*` and `uint8_t*` are interchangeable.

> 2. If not, should `kms_request_str_t::str` have type `uint8_t*`?

I think not. `kms_request_str` is a string wrapper type. Much of the API (`kms_request_strdupf`, `kms_request_str_appendf`, `kms_request_str_append_chars`, `kms_request_str_wrap`, `kms_request_str_new_from_chars`) accept `char *`.

I used `kms_request_str` in `kms_kmip_response_parser` as a data buffer. It was a shortcut to avoid defining a new buffer type.

If swapping between `uint8_t*` and `char*` is a concern, I am happy to file a MONGOCRYPT ticket to create a `kms_buffer_t` supporting an append operation to use in `kms_kmip_response_parser`. But I think `kms_request_str_t::str` should remain a `char *`.

Reply:
The cast itself is unlikely to be problematic for reasons you stated. However, I believe it preferable to avoid using `char*` to denote a buffer of bytes, as `char*` is already heavily overloaded in meaning. This appears to be the case here as well, where a logical "string" type `kms_request_str_t` is being used as byte storage instead. The need to cast between `uint8_t <-> char` is indicative of this logical discrepancy imo. However, I may be excessively applying C++ type safety idioms here; if repurposing `kms_request_str` as byte storage is considered acceptable, then I have no further issues with the proposed design.

Reply:
I think it is acceptable for now, since it does not affect correctness on platforms we support. But I do not think it is ideal. `kms_request_str` also NULL terminates, which is not applicable. I filed MONGOCRYPT-349 to make a future change to not use `kms_request_str_t`.